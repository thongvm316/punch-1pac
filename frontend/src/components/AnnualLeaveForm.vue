<template>
  <div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.attendance_day.$error || errors.attendance_day }"
    >
      <label class="form-label">{{ $t('label.annualLeaveDay') }}</label>
      <flat-pickr
        v-model="$v.params.attendance_day.$model"
        :config="{mode: 'single', locale: flatpickrLocaleMapper[pickrLocale]}"
        class="form-input daterange-picker"
      />
      <p
        v-if="$v.params.attendance_day.$error"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.annualLeaveDay') }) }}
      </p>
      <p
        v-if="errors.attendance_day"
        class="form-input-hint"
      >
        {{ $t('label.annualLeaveDay') }} {{ errors.attendance_day[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.reason.$error || errors.reason }"
    >
      <label class="form-label">{{ $t('label.reason') }}</label>
      <textarea
        v-model="$v.params.reason.$model"
        class="form-input"
      />
      <p
        v-if="$v.params.reason.$error"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.reason') }) }}
      </p>
      <p
        v-if="errors.reason"
        class="form-input-hint"
      >
        {{ $t('label.reason') }} {{ errors.reason[0] }}
      </p>
    </div>
    <div class="form-group">
      <button
        v-if="!request"
        ref="createAnnualLeaveBtn"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisable"
        @click="create()"
      >
        {{ $t('button.common.submit') }}
      </button>
      <button
        v-if="request"
        ref="updateAnnualLeaveBtn"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisable"
        @click="update()"
      >
        {{ $t('button.common.save') }}
      </button>
    </div>
  </div>
</template>

<script>
import { mapActions, mapState, mapMutations } from 'vuex'
import flatpickrLocale from '@/mixins/flatpickr-locale'
import handleSuccess from '@/mixins/handle-success'
import annualLeaveValidate from '@/validations/annual-leave-validate'
import { CLEAR_REQUEST_ERRORS } from '@/store/mutation-types'
const flatPickr = () => import('vue-flatpickr-component')

export default {
  name: 'AnnualLeaveForm',

  components: {
    flatPickr
  },

  mixins: [flatpickrLocale, handleSuccess, annualLeaveValidate],

  props: {
    request: {
      type: Object,
      default: null
    },

    type: {
      type: String,
      default: ''
    },

    annualDay: {
      type: String,
      default: ''
    }
  },

  data() {
    return {
      params: {
        attendance_day: '',
        reason: ''
      }
    }
  },

  computed: {
    ...mapState('requests', ['errors']),

    dataRequest() {
      return Object.assign(this.params, { kind: 'annual_leave' })
    }
  },

  created() {
    if (this.request) this.params = { ...this.request }
    if (this.annualDay) this.params.attendance_day = this.annualDay
  },

  methods: {
    ...mapMutations('requests', ['CLEAR_REQUEST_ERRORS']),
    ...mapActions('requests', ['addRequest', 'updateRequest']),

    create() {
      this.addRequest(this.dataRequest)
        .then(response => {
          this.handleSuccess({
            emitType: 'finishRequest',
            message: this.$t('messages.request.createSuccess')
          })

          this[CLEAR_REQUEST_ERRORS]()
        })
    },

    update() {
      const request = {
        id: this.request.id,
        params: this.dataRequest
      }

      this.updateRequest(request)
        .then(response => {
          this.handleSuccess({
            emitType: 'finishRequest',
            message: this.$t('messages.request.updateSuccess')
          })

          this[CLEAR_REQUEST_ERRORS]()
        })
    }
  }
}
</script>
