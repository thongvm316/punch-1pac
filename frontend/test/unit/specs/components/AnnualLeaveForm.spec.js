import { shallowMount } from '@vue/test-utils'
import setComputed from '../../supports/set-computed'
import requestsData from '../../supports/fixtures/report.api'
import wrapperOps from '../../supports/wrapper'
import AnnualLeaveForm from '@/components/AnnualLeaveForm'

const create = jest.spyOn(AnnualLeaveForm.methods, 'create')
const update = jest.spyOn(AnnualLeaveForm.methods, 'update')
const addRequest = jest.spyOn(AnnualLeaveForm.methods, 'addRequest').mockResolvedValue(null)
const updateRequest = jest.spyOn(AnnualLeaveForm.methods, 'updateRequest').mockResolvedValue(null)
const handleSuccess = jest.spyOn(AnnualLeaveForm.mixins[1].methods, 'handleSuccess')
const CLEAR_REQUEST_ERRORS = jest.spyOn(AnnualLeaveForm.methods, 'CLEAR_REQUEST_ERRORS')

Object.assign(wrapperOps, {
  stubs: {
    flatPickr: true
  }
})

describe('AnnualLeaveForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(AnnualLeaveForm, wrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when AnnualLeaveForm was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when error', () => {
    it('should display error attendance_day', async () => {
      wrapper.setData({ errors: { attendance_day: ['something wrong'] } })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })

    it('should display error reason', async () => {
      wrapper.setData({ errors: { reason: ['error'] } })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })

    it('should display both errors', async () => {
      wrapper.setData({ errors: { attendance_day: ['attendance_day error'], reason: ['reason error'] } })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when create request', () => {
    it('should call create method', async () => {
      setComputed(wrapper, {
        dataRequest: requestsData[0],
        isDisable: false
      })
      wrapper.find({ ref: 'createAnnualLeaveBtn' }).trigger('click')
      await wrapper.vm.$nextTick

      expect(create).toHaveBeenCalled()
      expect(addRequest).toHaveBeenCalledWith(requestsData[0])
      expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'finishRequest', message: 'Request is created' })
      expect(CLEAR_REQUEST_ERRORS).toHaveBeenCalled()
    })
  })

  describe('when update request', () => {
    beforeEach(() => {
      wrapper.setProps({ request: { id: 0, reason: 'personal issue' } })
      setComputed(wrapper, {
        isDisable: false
      })
    })

    it('should render update button', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call update method', async () => {
      wrapper.find({ ref: 'updateAnnualLeaveBtn' }).trigger('click')
      await wrapper.vm.$nextTick

      expect(update).toHaveBeenCalled()
      expect(updateRequest).toHaveBeenCalledWith({ id: 0, params: { attendance_day: '', kind: 'annual_leave', reason: '' } })
      expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'finishRequest', message: 'Request is updated' })
      expect(CLEAR_REQUEST_ERRORS).toHaveBeenCalled()
    })
  })
})
