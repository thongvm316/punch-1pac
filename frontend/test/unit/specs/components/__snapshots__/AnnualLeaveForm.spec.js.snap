// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnualLeaveForm.vue when AnnualLeaveForm was mounted should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Day Off</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" disabled="disabled" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`AnnualLeaveForm.vue when error should display both errors 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Day Off</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <p class="form-input-hint">
      Day Off attendance_day error
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <p class="form-input-hint">
      Reason reason error
    </p>
  </div>
  <div class="form-group"><button type="button" disabled="disabled" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`AnnualLeaveForm.vue when error should display error attendance_day 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Day Off</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <p class="form-input-hint">
      Day Off something wrong
    </p>
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" disabled="disabled" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`AnnualLeaveForm.vue when error should display error reason 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Day Off</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <p class="form-input-hint">
      Day Off something wrong
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <p class="form-input-hint">
      Reason error
    </p>
  </div>
  <div class="form-group"><button type="button" disabled="disabled" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`AnnualLeaveForm.vue when update request should render update button 1`] = `
<div>
  <div class="form-group"><label class="form-label">Day Off</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group">
    <!----> <button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;
